x-pack/plugins/alerting/server/saved_objects/migrations.ts

```diff
@@ -574,6 +589,49 @@ function removeMalformedExceptionsList(
   }
 }
 
+/**
+ * This migrates rule_id's within the legacy siem.notification to saved object references on an upgrade.
+ * We only migrate if we find these conditions:
+ *   - ruleAlertId is a string and not null, undefined, or malformed data.
+ *   - The existing references do not already have a ruleAlertId found within it.
+ * Some of these issues could crop up during either user manual errors of modifying things, earlier migration
+ * issues, etc... so we are safer to check them as possibilities
+ * @deprecated Once we are confident all rules relying on side-car actions SO's have been migrated to SO references we should remove this function
+ * @param doc The document that might have ruleId's to migrate into the references
+ * @returns The document migrated with saved object references
+ */
+function addRuleIdsToLegacyNotificationReferences(
+  doc: SavedObjectUnsanitizedDoc<RawAlert>
+): SavedObjectUnsanitizedDoc<RawAlert> {
+  const {
+    attributes: {
+      params: { ruleAlertId },
+    },
+    references,
+  } = doc;
+  if (!isSecuritySolutionLegacyNotification(doc) || !isString(ruleAlertId)) {
+    // early return if we are not a string or if we are not a security solution notification saved object.
+    return doc;
+  } else {
+    const existingReferences = references ?? [];
+    const existingReferenceFound = existingReferences.find((reference) => {
+      return reference.id === ruleAlertId && reference.type === 'rule';
```

Comment:
The rule saved object still has the type `alert` (confusing I know!) so I believe this should be checking for `alert` and any function that is extracting references from this saved object should be using type `alert`

Reply:
Thanks super helpful. I will change all the locations of this to be of type 'alert' and then the naming will be `param:alert_0` everywhere.